Add new console/http loggers

vadmeste opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
- Add console target logging, enabled by default.
- Add http target logging, which supports an endpoint
  with basic authentication (username/password are passed
  in the endpoint url itself)
- HTTP target logging is asynchronous and some logs can be
  dropped if channel buffer (1000 entries) is full

## Motivation and Context
Implement console/http loggers

## How Has This Been Tested?
Manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on cmd/logger/http.go:
```diff
@@ -0,0 +1,80 @@
+			if err != nil {
+				continue
+			}
+			defer resp.Body.Close()
```
You should close the body right here not in `defer` otherwise this will leak the connections.

vadmeste replied:
> You should close the body right here not in defer otherwise this will leak the connections.

Good catch, fixed

harshavardhana commented on cmd/logger/http.go:
```diff
@@ -0,0 +1,80 @@
+
+	// Create a routine which sends json logs received
+	// from an internal channel.
+	go func() {
```
This can move to a function something like h.startHTTPLogger() ?

vadmeste replied:
> This can move to a function something like h.startHTTPLogger() ?

Done

harshavardhana commented on cmd/logger/http.go:
```diff
@@ -0,0 +1,80 @@
+func NewHTTP(endpoint string) LoggingTarget {
+	h := HTTPTarget{
+		endpoint: endpoint,
+		logCh:    make(chan logEntry, 1000),
```
A 1000 is not enough perhaps we should keep a higher value?

vadmeste replied:
> A 1000 is not enough perhaps we should keep a higher value?

I increased it to 10k.

harshavardhana commented on cmd/logger/http.go:
```diff
@@ -0,0 +1,80 @@
+				continue
+			}
+
+			req, err := http.NewRequest("POST", h.endpoint, bytes.NewBuffer(logJSON))
```
Better to use our NewCustomHTTPTransport here since we might be leaving a connections in time_wait

vadmeste replied:
> Better to use our NewCustomHTTPTransport here since we might be leaving a connections in time_wait

Done.

donatello replied:
We should re-use the connection across log messages for better efficiency.

donatello commented on cmd/config-current.go:
```diff
@@ -232,6 +234,14 @@ func newServerConfig() *serverConfig {
+
+	// Create an example of HTTP logger
+	srvCfg.Logger.HTTP = make(map[string]loggerHTTP)
+	srvCfg.Logger.HTTP["target1"] = loggerHTTP{Endpoint: "https://login:password@endpoint.com/api"}
```
We could use a more standard example-ly link:

```
srvCfg.Logger.HTTP["target1"] = loggerHTTP{Endpoint: "https://username:password@example.com/api"}
```

vadmeste replied:
> srvCfg.Logger.HTTP["target1"] = loggerHTTP{Endpoint: "https://username:password@example.com/api"}

Done

donatello commented on cmd/logger/http.go:
```diff
@@ -0,0 +1,89 @@
+	h := HTTPTarget{
+		endpoint:  endpoint,
+		transport: transport,
+		logCh:     make(chan logEntry, 10000),
```
If we place the `http.Client` here, and re-use, I think connection re-use would be automatic?

vadmeste replied:
> We should re-use the connection across log messages for better efficiency.

Connection reusing means establishing a permanent connection with the server ? if yes, then I don't recommend it since an established connection will consume one tcp port meanwhile servers are not supposed to send logs regularly. 

What do you think ?

donatello replied:
The idle connection timeout of 30 seconds takes care of this. If multiple logs were generated, then repeatedly opening TCP connections, and negotiating TLS handshake, etc is too repetitive - it is better to re-use an established connection.

vadmeste replied:
> The idle connection timeout of 30 seconds takes care of this. If multiple logs were generated, then repeatedly opening TCP connections, and negotiating TLS handshake, etc is too repetitive - it is better to re-use an established connection.

Yeah, didn't know about that 30 seconds timeout, fixed now

vadmeste committed:
Add new console/http loggers

- Add console target logging, enabled by default.
- Add http target logging, which supports an endpoint
  with basic authentication (username/password are passed
  in the endpoint url itself)
- HTTP target logging is asynchronous and some logs can be
  dropped if channel buffer (10000) is full

kannappanr reviewed (approved):
tested and LGTM
